Allow display of unplaced labels

nyalldawson opened the pull request:
This setting (accessed through the project label settings dialog or via an action on the labeling toolbar) allows unplaced labels to be shown on the map. These are rendered using a red color by default (but the color can be changed from the label settings dialog too), allowing users to determine whether any important labels are missing from their maps (e.g. due to overlaps or other constraints).

Additionally, there's a similar checkbox present in the labeling panel for a layout map, allowing unplaced labels in that particular layout map to be displayed.

nyalldawson committed:
Append additional "QgsCalloutContext" member to callout rendering virtual
methods

While unused for now, this gives us flexibility in future to specify
additional useful contextual information about how a callout should
be rendered without breaking API (e.g. label text alignment, label
font settings, etc)

nyalldawson committed:
Add API to pal to return unplaced labels after calculating the solution
to the labeling problem

And improve dox

nyalldawson committed:
[FEATURE][labeling] Allow display of unplaced labels

This setting (accessed through the project label settings dialog)
allows unplaced labels to be shown on the map. These are rendered
using a red color by default (but the color can be changed from
the label settings dialog too), allowing users to determine whether
any important labels are missing from their maps (e.g. due to
overlaps or other constraints).

nyalldawson committed:
[needs-docs] Add labeling toolbar action to toggle whether unplaced
labels are displayed

nyalldawson committed:
[needs-docs] Add checkbox to layout map label settings panel to allow
display of unplaced labels

This is independant from the main canvas setting, and can be used
to determine whether labels are missing from a particular layout map

nyalldawson committed:
Prioritize moving unplaced labels in the move label tool

These are most likely the ones which users are trying to move
when there's overlapping labels

nyalldawson commented:
@nirvn needing your icon design skills for a proper toolbar icon here!

nirvn commented on src/app/layout/qgslayoutmapwidget.cpp:
```diff
@@ -1601,6 +1601,7 @@ QgsLayoutMapLabelingWidget::QgsLayoutMapLabelingWidget( QgsLayoutItemMap *map )
   connect( mLabelBoundaryUnitsCombo, &QgsLayoutUnitsComboBox::changed, this, &QgsLayoutMapLabelingWidget::labelMarginUnitsChanged );
   connect( mLabelBoundarySpinBox, static_cast < void ( QDoubleSpinBox::* )( double ) > ( &QDoubleSpinBox::valueChanged ), this, &QgsLayoutMapLabelingWidget::labelMarginChanged );
   connect( mShowPartialLabelsCheckBox, &QCheckBox::toggled, this, &QgsLayoutMapLabelingWidget::showPartialsToggled );
+  connect( mDrawUnplacedCheckBox, &QCheckBox::toggled, this, &QgsLayoutMapLabelingWidget::showUnplacedToggled );
```
Cosmetic suggestion re checkbox and widget function name: settle on draw vs. show. I'd suggest sticking to show.

nirvn commented:
@nyalldawson , here's an icon:
[mActionShowUnplacedLabel.zip](https://github.com/qgis/QGIS/files/3413307/mActionShowUnplacedLabel.zip)

nyalldawson committed:
Also draw unlabel features which correspond to features where no
candidates could be generated

E.g. lines too short for label, polygons too small for labels

(and fix some gross oldschool memory management)

nyalldawson committed:
Use dedicated icon for 'show unplaced'

nyalldawson committed:
Move show unplaced icon to near start of labeling toolbar

nyalldawson committed:
Use show instead of draw

nyalldawson committed:
Add test for drawing unplaced for a feature with no candidates

nyalldawson commented:
Thanks @nirvn, integrated now

nirvn commented:
@nyalldawson , code looking good. +1 to merge when travis turns green.

nyalldawson commented:
Thanks! Let qgis 3.10 forever be known as the "labelling revolution"
